File: dagger2-scopes-and-subcomponents-part1/app/src/main/java/com/example/dagger2scopeandsubcomponentspart1/AppModule.java
package com.example.dagger2scopeandsubcomponentspart1;

import dagger.Module;
import dagger.Provides;

@Module
public class AppModule {
    private int index = 0;

    @Provides
    public Warrior provideWarrior() {
        index++;
        return new Warrior("Warrior " + index);
    }
}


//example with applicationscope
//@Module
//public class AppModule {
//    private int index = 0;
//
//    @Provides
//    @ApplicationScope
//    public Warrior provideWarrior() {
//        index++;
//        return new Warrior("Warrior " + index);
//    }
//}

File: dagger2-scopes-and-subcomponents2/app/src/main/java/com/example/dagger2scopesandsubcomponentspart2/AppModule.kt
package com.example.dagger2scopesandsubcomponentspart2

import dagger.Module
import dagger.Provides

@Module
class AppModule {
    private var index = 0

    @ApplicationScope
    @Provides
    fun provideWarrior(): Warrior {
        index++
        return Warrior("Warrior $index")
    }
}